### PR TITLE
Feature: Refactor modal types and introduce side sheet 

### PR DESCRIPTION
**Pull request by ulusoyca**
## Description

This PR replaces the `WoltModalType` enum with the `WoltModalType` abstract class, designed to standardize the behavior and appearance of various modal types across the application. It also includes the implementation of specific modal types that extend WoltModalType, such as WoltDialogType, WoltBottomSheetType, and WoltSideSheetType. This implementation ensures consistency in our modals' design and enables modal type customization.

> [!NOTE]  
> This PR updates the existing specs of dialog and bottom sheets according to Wolt's design system, and introduces side sheet specs.

Here is the specs:
alc_design_break_points

### Screenshots with safe area

| Verical Bottomsheet | Horizontal Bottomsheet |
|--------|--------|
| image | image |
| image | image | 

| Verical Dialog | Horizontal Dialog |
|--------|--------|
| image | image |
| image | image | 

| Vertical Side Sheet | Horizontal Side Sheet |
|--------|--------|
| image | image |
| image | image | 

| LTR Side Sheet | LTR SIde Sheet |
|--------|--------|
| image | image |
| image | image | 

| Bottomsheet Responsive |
|--------|
| <video src="https://github.com/woltapp/wolt_modal_sheet/assets/13782003/5a4f62d5-89ad-4426-b990-709b7ec77b17"> |

| Dialog Responsive |
|--------|
| <video src="https://github.com/woltapp/wolt_modal_sheet/assets/13782003/675fb345-b9e3-42c5-934e-247c10f14b91"> |

| Side sheet responsive |
|--------|
| <video src="https://github.com/woltapp/wolt_modal_sheet/assets/13782003/d7c5ecbd-08bf-435e-9129-ac729fdb1505"> |

## Issues Related
Issues related to this feature is labeled with [Custom Modal Type](https://github.com/woltapp/wolt_modal_sheet/issues?q=is%3Aissue+is%3Aopen+label%3A%22Custom+Modal+Type%22)
- https://github.com/woltapp/wolt_modal_sheet/issues/229
- https://github.com/woltapp/wolt_modal_sheet/issues/170
- https://github.com/woltapp/wolt_modal_sheet/issues/25
- https://github.com/woltapp/wolt_modal_sheet/issues/224

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors.
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] The package compiles with the minimum Flutter version stated in the [pubspec.yaml](https://github.com/woltapp/wolt_modal_sheet/blob/main/pubspec.yaml#L8)

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change.
- [ ] No, this is *not* a breaking change.

[Contributor Guide]: https://github.com/woltapp/wolt_modal_sheet/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Review by yunusemrebakir — APPROVED** [review]

File: lib/src/wolt_modal_sheet_route.dart [hunk 1]
```diff
@@ -143,37 +107,29 @@ class WoltModalSheetRoute<T> extends PageRoute<T> {
...
 
   @override
-  Duration get transitionDuration => _transitionDuration;
+  Duration get transitionDuration {
+    return _determineCurrentModalType(navigator!.context).transitionDuration;
+  }
+
+  @override
+  Duration get reverseTransitionDuration =>
+      _determineCurrentModalType(navigator!.context).reverseTransitionDuration;
```
Comment (TahaTesser) [inline lib/src/wolt_modal_sheet_route.dart:132]: Minor nit: `transitionDuration` uses curly braces and `reverseTransitionDuration` uses return operator. The package has been using return operator so we should keep it consistent.

**Review by TahaTesser — APPROVED** [review]
This is mind blowing. Excellent work1

Apologies for delayed review as I've been moving to another city this.

File: lib/src/content/components/main_content/wolt_modal_sheet_main_content.dart [hunk 2]
```diff
@@ -87,7 +87,8 @@ class WoltModalSheetMainContent extends StatelessWidget {
           ...page.mainContentSlivers!
         else
           ...page.mainContentSliversBuilder!(context),
-        if (page.forceMaxHeight && !isNonScrollingPage)
+        if (woltModalType.forceMaxHeight ||
```
[the remark this answers is not in the document]
Comment (ulusoyca) [reply lib/src/content/components/main_content/wolt_modal_sheet_main_content.dart]: here we go: d986dcad289af76be848d33aa9ec22ea215aedc5

File: lib/src/wolt_modal_sheet_route.dart (hunk 1 above)
Comment (ulusoyca) [reply lib/src/wolt_modal_sheet_route.dart]: done ffc63266f27c38e080f613f4d6361fa3c2cf1010

File: lib/src/theme/wolt_modal_sheet_default_theme_data.dart [hunk 3]
```diff
@@ -132,28 +110,27 @@ class WoltModalSheetDefaultThemeData extends WoltModalSheetThemeData {
...
+  bool get useSafeArea => true;
+
+  @override
+  WoltModalTypeBuilder get modalTypeBuilder => (context) {
+        final width = MediaQuery.sizeOf(context).width;
+        if (width < 524.0) {
+          return WoltModalType.bottomSheet();
+        } else {
+          return WoltModalType.dialog();
+        }
```
[the remark this answers is not in the document]
Comment (ulusoyca) [reply lib/src/theme/wolt_modal_sheet_default_theme_data.dart]: done ffc63266f27c38e080f613f4d6361fa3c2cf1010

File: lib/src/modal_type/wolt_dialog_type.dart [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        maxHeight: max(360, availableHeight * 0.8),
+      );
+    } else {
+      return BoxConstraints(
+        minWidth: width,
+        maxWidth: width,
+        minHeight: availableHeight * 0.8,
+        maxHeight: availableHeight * 0.8,
+      );
+    }
```
[the remark this answers is not in the document]
Comment (ulusoyca) [reply lib/src/modal_type/wolt_dialog_type.dart]: done ffc63266f27c38e080f613f4d6361fa3c2cf1010

File: lib/src/modal_type/wolt_dialog_type.dart [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+  /// [availableSize] is the total available space for the modal.
+  ///
+  /// Returns [BoxConstraints] that define the maximum and minimum dimensions of the dialog.
+  @override
+  BoxConstraints layoutModal(Size availableSize) {
+    const padding = 48.0;
+    double width;
+    final availableWidth = availableSize.width;
+    const verticalBreakpoint = 360.0 + padding;
+    const horizontalBreakpoint = 524.0 + padding;
```
[the remark this answers is not in the document]
Comment (ulusoyca) [reply lib/src/modal_type/wolt_dialog_type.dart]: done ffc63266f27c38e080f613f4d6361fa3c2cf1010

**Pull request merged**